sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java

```diff
@@ -3729,6 +3729,67 @@ public void testTwoTimersSettingEachOtherWithCreateAsInput() {
       pipeline.run();
     }
 
+    @Test
+    @Category({
+      ValidatesRunner.class,
+      UsesStatefulParDo.class,
+      UsesTimersInParDo.class,
+      UsesTestStream.class
+    })
+    public void testValueStateSimple() {
+      final String stateId = "foo";
+      final String timerId = "bar";
+      DoFn<KV<String, Integer>, Integer> fn =
+          new DoFn<KV<String, Integer>, Integer>() {
+
+            @TimerId(timerId)
+            private final TimerSpec timer = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+            @ProcessElement
+            public void processElement(ProcessContext c, @TimerId(timerId) Timer timer) {
+              timer.withOutputTimestamp(new Instant(5)).set(new Instant(8));
+            }
+
+            @OnTimer(timerId)
+            public void onTimer(OnTimerContext c, BoundedWindow w) {
+              c.output(100);
+            }
+          };
+
+      DoFn<Integer, Instant> fn1 =
+          new DoFn<Integer, Instant>() {
+
+            @ProcessElement
+            public void processElement(ProcessContext c) {
+              c.output(c.timestamp());
+            }
+          };
+
+      Instant base = new Instant(0);
+
+      TestStream<KV<String, Integer>> stream =
+          TestStream.create(KvCoder.of(StringUtf8Coder.of(), VarIntCoder.of()))
+              .advanceWatermarkTo(new Instant(0))
+              .addElements(KV.of("key", 1))
+              .advanceWatermarkTo(new Instant(100))
+              .advanceWatermarkToInfinity();
```

Comment:
I don't think that this test quite tests the watermark holds.

Here's a structure that I think would test what we want:

doFn1: receives element and sets a timer for 100 with output timestamp of 5. Also outputs that element downstream
doFn2: receives element from doFn1 and sets a timer for 50.

Now if the input watermark is advanced to 70, we expect nothing to happen. If there were no watermark holds, then doFn2's timer would fire (because doFn1 would be only holding the watermark to 100). 

To test this I think we should advance the input watermark to 70, and then send a new input element to the pipeline. Upon receiving this special input, dofn1 should forward it to dofn2, and dofn2 should verify that the timer has not yet fired (it should do this by storing  in ValueState whether the timer has fired already). Then when we advance the watermark to 100 we expect both timers to fire.

Reply:
Let me know if there are more questions on this.

Reply:
Actually a simpler way to test this: if you remove the withAllowedLateness call here I think you will also be testing this. Because if the watermark were not help up, it would be late data and dropped.

Reply:
It looks like we're still working on this test, correct?

Reply:
We have removed with AllowedLateness call here as per your suggestion. I think we are done with the test.

Reply:
@kennknowles do you know when allowedLateness elements are dropped? In this case there is no downstream ParDo (just captured by PAssert) - should we still expect the elements to be dropped if there is no downstream shuffle or ParDo? I'm not 100% convinced yet that this test is working as expected.

Reply:
Also withAllowedLateness is still being specified below.